tyrian/js/src/main/scala/tyrian/websocket/WebSocket.scala

```diff
@@ -57,104 +58,105 @@ object KeepAliveSettings:
 
   private def connectTask[F[_]: Async](
       address: String,
       onOpenMessage: String,
-      keepAliveSettings: KeepAliveSettings
+      keepAliveSettings: KeepAliveSettings,
+      disp: Dispatcher[F]
   ): F[WebSocketConnect[F]] =
-    Async[F].map(newConnection(address, Some(onOpenMessage), keepAliveSettings))(ls =>
+    Async[F].map(newConnection(address, Some(onOpenMessage), keepAliveSettings, disp))(ls =>
       WebSocketConnect.Socket(WebSocket(ls))
     )
 
   private def newConnection[F[_]: Async](
       address: String,
       onOpenSendMessage: Option[String],
-      settings: KeepAliveSettings
+      settings: KeepAliveSettings,
+      disp: Dispatcher[F]
   ): F[LiveSocket[F]] =
-    Async[F].delay {
-      val socket    = new dom.WebSocket(address)
-      val keepAlive = new KeepAlive(socket, settings)
-      val q         = Queue.empty[WebSocketEvent]
-
-      val msgListener = Functions.fun { e =>
-        val event = WebSocketEvent.Receive(e.asInstanceOf[dom.MessageEvent].data.toString)
-        q.enqueue(event)
-      }
-
-      val errListener = Functions.fun { e =>
-        val msg =
-          try e.asInstanceOf[dom.ErrorEvent].message
-          catch { case _: Throwable => "Unknown" }
-        q.enqueue(WebSocketEvent.Error(msg))
-      }
-
-      val closeListener = Functions.fun { e =>
-        val ev = e.asInstanceOf[dom.CloseEvent]
-        q.enqueue(WebSocketEvent.Close(ev.code, ev.reason))
-      }
+    // XXX - How to get flatMap into scope for a for-comprehension?
+    Async[F].flatMap(Queue.unbounded[F, WebSocketEvent]) { q =>
+      Async[F].delay {
+        val socket    = new dom.WebSocket(address)
+        val keepAlive = new KeepAlive(socket, settings)
+
+        val msgListener = Functions.fun { e =>
+          val event = WebSocketEvent.Receive(e.asInstanceOf[dom.MessageEvent].data.toString)
+          disp.unsafeRunAndForget(q.offer(event))
+        }
+
+        val errListener = Functions.fun { e =>
+          val msg =
+            try e.asInstanceOf[dom.ErrorEvent].message
+            catch { case _: Throwable => "Unknown" }
+          disp.unsafeRunAndForget(q.offer(WebSocketEvent.Error(msg)))
+        }
+
+        val closeListener = Functions.fun { e =>
+          val ev = e.asInstanceOf[dom.CloseEvent]
+          disp.unsafeRunAndForget(q.offer(WebSocketEvent.Close(ev.code, ev.reason)))
+        }
+
+        val openListener = Functions.fun { _ =>
+          onOpenSendMessage.foreach(msg => socket.send(msg))
+          disp.unsafeRunAndForget(q.offer(WebSocketEvent.Open))
+        }
+
+        socket.addEventListener("message", msgListener)
+        socket.addEventListener("error", errListener)
+        socket.addEventListener("open", openListener)
+        socket.addEventListener("close", closeListener)
+
+        val subs =
+          Sub.Batch(
+            Sub.make(s"[tyrian-ws-${address}]")(
+              Stream.repeatEval {
```

Comment:
So what we're saying is that although we're still using `repeatEval`, `take` will block fiber when it's empty, effectively preventing a busy loop - is that about right?

Reply:
Actually this can just be `Stream.fromQueueUnterminated(q)` 🙁 

You might also look at `Channel` instead, it is optimized for this specific use-case.
https://github.com/typelevel/fs2/blob/509d74fdcc5bf0735cabcf829d0a5ba21a949ed9/core/shared/src/main/scala/fs2/concurrent/Channel.scala#L29-L31

Reply:
@davesmith00000 - Yes, that's exactly right.

@armanbilge - Oh, cool! Thank you for the pointer. I'll take a look.